### PR TITLE
Add solver for linear and least square to linalg

### PR DESCRIPTION
**Pull request by yingryic**
corresponds to #2526, built interface to call Eigen3 internal dense solver for linear systems (LLT) and least square problem (Householder QR). Please check if this form is what you want, @karlnapf May add more solver's later.

### CODE REVIEW

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+CDirectDenseLeastSquareSolverQR::~CDirectDenseLeastSquareSolverQR()
+{
+}
+
+SGVector<float64_t> CDirectDenseLeastSquareSolverQR::solve(
+		CLinearOperator<SGVector<float64_t>, SGVector<float64_t> >* A, SGVector<float64_t> b)
+{
+	REQUIRE(A, "Operator is NULL!\n");
+	REQUIRE(A->get_dimension() == b.vlen, "Dimension mismatch!\n");
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.cpp]: "Matrix dimension (%d) does not match vector dimension (%d)\n"

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	CDenseMatrixOperator<float64_t>* op
+		=dynamic_cast<CDenseMatrixOperator<float64_t>*>(A);
+	REQUIRE(op, "Operator is not DenseMatrixOperator type!\n");
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.cpp]: "Operator \"%s\" is not of type DenseMatrixOperator.\n", op->get_name()

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#include <shogun/lib/config.h>
+
+#ifdef HAVE_EIGEN3
+#include <shogun/mathematics/linalg/linsolver/LinearSolver.h>
+
+namespace shogun
+{
+
+/** @brief Class that provides a solve method for least square problem
+ * using QR decomposition
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.h]: Some math details would be good, as well as a wikipedia link.
Also, state the assumptions to the operator

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+{
+public:
+	/** default constructor */
+	CDirectDenseLeastSquareSolverQR();
+
+	/** destructor */
+	virtual ~CDirectDenseLeastSquareSolverQR();
+
+	/**
+	 * solve method for solving real-valued least square problem based on QR decomposition
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.h]: Capital letters in sentence starts in all exposed documentation  please

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLeastSquareSolverQR.cpp]: We changed to BSD, see other patches

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.cpp]: BSD

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+CDirectDenseLinearSolverLLT::~CDirectDenseLinearSolverLLT()
+{
+}
+
+SGVector<float64_t> CDirectDenseLinearSolverLLT::solve(
+		CLinearOperator<SGVector<float64_t>, SGVector<float64_t> >* A, SGVector<float64_t> b)
+{
+	REQUIRE(A, "Operator is NULL!\n");
+	REQUIRE(A->get_dimension()==b.vlen, "Dimension mismatch!\n");
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.cpp]: same comments as above

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	Map<VectorXd> map_b(b.vector, b.vlen);
+	
+	// using LLT to solve the system Ax=b
+	LLT<MatrixXd> llt;
+	llt.compute(map_m);
+	map_x=llt.solve(map_b);
+	
+	// checking for success
+	if (llt.info()!=Success)
+		SG_WARNING("LLU solver failed! maybe input operator is not symmetric positive definite.\n");
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.cpp]: Maybe with capital.

LLT solver

Maybe? What *is* the reason? Can't we ask eigen3?

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.h [hunk 9]
```diff
@@ -0,0 +1,54 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General turalPublic License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Written (W) 2015 Yingrui Chang
+ */
+
+#ifndef DIRECT_DENSE_LINEAR_SOLVER_LLT_H_
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.h]: please no _ at the end

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.h [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+#include <shogun/lib/config.h>
+
+#ifdef HAVE_EIGEN3
+#include <shogun/mathematics/linalg/linsolver/LinearSolver.h>
+
+namespace shogun
+{
+
+/** @brief Class that provides a solve method for real SPD dense systems
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.h]: math? wikipedia?

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.h [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+{
+public:
+	/** default constructor */
+	CDirectDenseLinearSolverLLT();
+
+	/** destructor */
+	virtual ~CDirectDenseLinearSolverLLT();
+
+	/**
+	 * solve method for solving real-valued SPD linear systems
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.h]: Solve

File: src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.h [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+/** @brief Class that provides a solve method for real SPD dense systems
+ * using LLT decomposition
+ */
+class CDirectDenseLinearSolverLLT : public CLinearSolver<float64_t, float64_t>
+{
+public:
+	/** default constructor */
+	CDirectDenseLinearSolverLLT();
+
+	/** destructor */
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/linsolver/DirectDenseLinearSolverLLT.h]: D

File: tests/unit/mathematics/linalg/DirectDenseLeastSquareSolverQR_unittest.cc [hunk 13]
```diff
@@ -0,0 +1,54 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
```
Comment (karlnapf) [inline tests/unit/mathematics/linalg/DirectDenseLeastSquareSolverQR_unittest.cc]: BSD

**Comment by karlnapf** [comment]
Great stuff!
This is definitely useful

@łambday should comment, too. I am surprised we did not add direct solvers back in the log-det project.

@yingryic You added the solvers to the linalg classes for linear operators. This is good and can be merged (modulo some style issues).

Now, we are interested in integrating the same type of thing into the linalg framework here:
https://github.com/shogun-toolbox/shogun/wiki/README_linalg

You will noticed that this is quite different, not based on implementing classes. But rather defining an interface (similar to the dot products we have) and the implement various backends for it. Let me know hwo this goes

**Comment by lambday** [comment]
Err I didn't notice this PR earlier. Thanks @karlnapf for notifying.

We *did* add direct linear solvers for linear operator interface back then during the logdet project. Please check [CDirectLinearSolverComplex](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/mathematics/linalg/linsolver/DirectLinearSolverComplex.cpp) which has LLT, QR, SVD. The only limitation is that it works with complex-valued linear operator and real-valued vector - which shouldn't be too hard to change if required.

I agree that it is good to have direct solvers for linop interfaces as well - but when we have the backend linsolvers ready, these classes get rid of eigen3 dependency. So I'd focus more on that as of now. 

@yingryic please ask away if you need any clarification regarding what we're planning to do here for those solvers. Check the README that Heiko shared, also have a look at the internal stuffs inside linalg.

**Comment by yingryic** [comment]
I made the style change accordingly, @karlnapf 

But it is indeed somewhat a repetition with the existing direct complex solver. I am totally fine if you choose not to merge it.

**Comment by karlnapf** [comment]
I dont know, wouldnt it be better to template the existing ones?
All this redundant code makes me nervous. If it is too messy to do (what do you think @lambday ) then we could have this one too. But the interfaces should be exactly the same pls.

**Comment by karlnapf** [comment]
I agree with lambday, we should have the interface in linalg, and then make these linear operators work on that interface.

@yingryic its good that you coded this up as you probably learned a lot. I suggest to keep this PR open for now and move to linalg? Then come back here?

**Comment by karlnapf** [comment]
I will close this for now, but it would be great if we had something similar for linalg finally

**Pull request closed without merging**